README.rst
================
django-cache-url
================

.. image:: https://img.shields.io/pypi/v/django-cache-url.svg
    :alt: PyPI version
    :target: https://pypi.org/project/django-cache-url/

.. image:: https://img.shields.io/pypi/pyversions/django-cache-url.svg
   :alt: Supported Python versions
   :target: https://pypi.org/project/django-cache-url/

.. image:: https://github.com/epicserve/django-cache-url/workflows/CI/badge.svg?branch=master
   :target: https://github.com/epicserve/django-cache-url/actions
   :alt: Tests

This simple Django utility allows you to utilize the
`12factor <http://www.12factor.net/backing-services>`_ inspired
``CACHE_URL`` environment variable to configure your Django application.

This was built with inspiration from rdegges'
`django-heroku-memcacheify <https://github.com/rdegges/django-heroku-memcacheify>`_
as a way to use CACHE_URL in apps that aren't necessarily hosted on Heroku.

The internals borrow heavily from kennethreitz's
`dj-database-url <https://github.com/kennethreitz/dj-database-url>`_.


Usage
-----
Configure your cache in ``settings.py`` from ``CACHE_URL``::

    import django_cache_url
    CACHES = {'default': django_cache_url.config()}

Defaults to local memory cache if ``CACHE_URL`` isn't set.

Parse an arbitrary Cache URL::

    CACHES = {'default': django_cache_url.parse('memcache://...')}

Supported Caches
----------------
Support currently exists for:

* locmem (default): ``'locmem://[NAME]'``
* db: ``'db://TABLE_NAME'``
* dummy: ``'dummy://'``
* file: ``'file:///PATH/TO/FILE'``
* memcached: ``'memcached://HOST:PORT'`` [#memcache]_
* pymemcached: ``'pymemcached://HOST:PORT'`` For use with the `python-memcached`_ library. Useful if you're using Ubuntu <= 10.04.
* pymemcache: ``'pymemcache://HOST:PORT'`` For use with the `pymemcache`_ library.
* djangopylibmc: ``'djangopylibmc://HOST:PORT'`` For use with SASL based setups such as Heroku.
* redis: ``'redis://[USER:PASSWORD@]HOST:PORT[/DB]'`` or ``'redis:///PATH/TO/SOCKET[/DB]'`` For use with `django-redis`_. To use `django-redis-cache`_ just add ``?lib=redis-cache`` to the URL.
* rediss: ``'rediss://[USER:PASSWORD@]HOST:PORT[/DB]'`` For use with `django-redis`_.
* hiredis: ``'hiredis://[USER:PASSWORD@]HOST:PORT[/DB]'`` or ``'hiredis:///PATH/TO/SOCKET[/DB]'`` For use with django-redis library using HiredisParser.
* hirediss: ``'hirediss://[USER:PASSWORD@]HOST:PORT[/DB]'`` or ``'hirediss:///PATH/TO/SOCKET[/DB]'`` For use with django-redis library using HiredisParser.
* uwsgicache: ``'uwsgicache://[CACHENAME]'`` For use with `django-uwsgi-cache`_. Fallbacks to ``locmem`` if not running on uWSGI server.

All cache urls support optional cache arguments by using a query string, e.g.: ``'memcached://HOST:PORT?key_prefix=site1'``. See the Django `cache arguments documentation`_.

.. [#memcache] To specify multiple instances, separate the the ``HOST:PORT``` pair
               by commas, e.g: ``'memcached://HOST1:PORT1,HOST2:PORT2``

.. _django-redis: https://github.com/niwibe/django-redis
.. _django-redis-cache: https://github.com/sebleier/django-redis-cache
.. _python-memcached: https://github.com/linsomniac/python-memcached
.. _pymemcache: https://github.com/pinterest/pymemcache
.. _cache arguments documentation: https://docs.djangoproject.com/en/dev/topics/cache/#cache-arguments
.. _django-uwsgi-cache: https://github.com/ionelmc/django-uwsgi-cache

Installation
------------
Installation is simple too::

    $ pip install django-cache-url

Tests
-----

To run the tests install tox::

    pip install tox

Then run them with::

    make test



CHANGELOG.rst
CHANGELOG
=========

v3.4.5
------

- Add support for Django 4.2, 5.0
- Add support for Python 3.11, 3.12
- Drop support for Python 3.6, 3.7
- Drop support for Django 2.2, 4.0, 4.1


v3.4.2
------

- Fix a bug where providing a password for the redis backend would produce an invalid configuration on Django 4.

v3.4.0
------

- Add support for Django 4.0

v3.3.0
------

- Add support for pymemcache.

v3.1.0
------

- Bring the project back into active status, so it can be used with `environs <https://github.com/sloria/environs>`_.
- Drop support for Python less than 3.8.

v3.0.0
------

- Deprecate project in favour of `Django Environ <https://pypi.org/project/django-environ/>`_.


v2.0.0
------

- **Backwards Incompatible** Remove Python 2.6 support
- Fix urls without a port getting their port set to "None" instead (thanks to Linus Lewandowski)


v1.4.0
------

- Add django-elasticache support (thanks to Alex Couper)


v1.3.1
------

- Fix django-redis support (thanks to Manatsawin Hanmongkolchai)


v1.3.0
------

- Support for django-redis >= 4.50 (thanks to Egor Yurtaev)


v1.2.0
------

- Run tests on Python 3.5 (thanks to Anton Egorov)
- Add support for MAX_ENTRIES and CULL_FREQUENCY options (thanks to Anton Egorov)


v1.1.0
------

- Add support for uwsgi caching (thanks to Alan Justino da Silva)


v1.0.0
------

- **Backwards Incompatible** Improve Redis URL parsing, making PREFIX parsing much easier and automatically converting query args into Django Cache settings (thanks to Russell Davies)
- **Backwards Incompatible** Switch to ``django-redis``'s new import name (thanks to Michael Warkentin)
- Switch to Tox for running tests with different pythons
- Switch to pytest


v0.8.0
------

- Add support for password in redis urls (thanks to Mjumbe Wawatu Ukweli)


v0.7.0
------

- Add support for UNIX sockets in redis urls (thanks to Jannis Leidel)


v0.6.0
------

- Fix Python 3 support


v0.5.0
------

- Add multiple memcache locations


v0.4.0
------

- Add redis and hiredis support


v0.3.4
------

- Fix Python 3 compatibility import bug


v0.3.3
------

- Add Python 3 compatibility


v0.3.2
------

- Fix setting name bug


v0.3.1
------

- Remove underscore from django pylibmc scheme


v0.3.0
------

- Add python memcached support
- Add django pylibmc support


v0.2.0
------

- Add prefix support
- Split up tests
- Tidy up examples


v0.1.0
------

- Initial release


tests/test_core.py
import os

import pytest

import django_cache_url


def test_config_defaults_to_locmem():
    if os.environ.get('CACHE_URL'):
        del os.environ['CACHE_URL']
    config = django_cache_url.config()
    assert config['BACKEND'] == 'django.core.cache.backends.locmem.LocMemCache'


def test_db_url_returns_database_cache_backend():
    url = 'db://super_caching_table'
    config = django_cache_url.parse(url)

    assert config['BACKEND'] == 'django.core.cache.backends.db.DatabaseCache'
    assert config['LOCATION'] == 'super_caching_table'


def test_dummy_url_returns_dummy_cache():
    config = django_cache_url.parse('dummy://')
    assert config['BACKEND'] == 'django.core.cache.backends.dummy.DummyCache'


def test_file_url_returns_file_cache_backend_relative():
    config = django_cache_url.parse('file:///herp/bar')

    assert config['BACKEND'] == 'django.core.cache.backends.filebased.FileBasedCache'
    assert config['LOCATION'] == 'herp/bar'


def test_file_url_returns_file_cache_backend_absolute():
    config = django_cache_url.parse('file:////var/herp')

    assert config['BACKEND'] == 'django.core.cache.backends.filebased.FileBasedCache'
    assert config['LOCATION'] == '/var/herp'


def test_locmem_url_returns_locmem_cache():
    config = django_cache_url.parse('locmem://')
    assert config['BACKEND'] == 'django.core.cache.backends.locmem.LocMemCache'


def test_query_string_params_are_converted_to_cache_arguments():
    url = 'redis:///path/to/socket?key_prefix=foo&bar=herp'
    config = django_cache_url.parse(url)

    assert config['KEY_PREFIX'] == 'foo'
    assert config['BAR'] == 'herp'


def test_query_string_params_are_converted_to_cache_options():
    url = 'db://my_cache_table?max_entries=1000&cull_frequency=2'
    config = django_cache_url.parse(url)

    assert 'OPTIONS' in config
    assert config['OPTIONS']['MAX_ENTRIES'] == 1000
    assert config['OPTIONS']['CULL_FREQUENCY'] == 2


def test_unknown_cache_backend():
    with pytest.raises(Exception):
        django_cache_url.parse('donkey://127.0.0.1/foo')


tests/test_redis.py

# region HIREDIS url
def test_hiredis_dj3(django_cache_url_dj3):
    url = 'hiredis://127.0.0.1:6379/0?key_prefix=site1'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://127.0.0.1:6379/0'
    assert config['OPTIONS']['PARSER_CLASS'] == 'redis.connection.HiredisParser'


def test_hiredis_dj4(django_cache_url_dj4):
    url = 'hiredis://127.0.0.1:6379/0?key_prefix=site1&pool_class=redis.BlockingConnectionPool'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://127.0.0.1:6379/0'
    assert 'PARSER_CLASS' not in config.get('OPTIONS', {})
    assert config['OPTIONS']['parser_class'] == 'redis.connection.HiredisParser'
    assert config['OPTIONS']['pool_class'] == 'redis.BlockingConnectionPool'
# endregion


# region HIREDIS socket
def test_hiredis_socket_dj3(django_cache_url_dj3):
    url = 'hiredis:///path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'unix:///path/to/socket?db=1'
    assert config['OPTIONS']['PARSER_CLASS'] == 'redis.connection.HiredisParser'


def test_hiredis_socket_dj4(django_cache_url_dj4):
    url = 'hiredis:///path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'unix:///path/to/socket?db=1'
    assert 'PARSER_CLASS' not in config.get('OPTIONS', {})
    assert config['OPTIONS']['parser_class'] == 'redis.connection.HiredisParser'


def test_hiredis_socket_with_username_password_dj3(django_cache_url_dj3):
    url = 'hiredis://foo:bar@/path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DJANGO_REDIS_BACKEND
    assert config['LOCATION'] == 'unix://foo:bar@/path/to/socket?db=1'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
    assert config['KEY_PREFIX'] == 'site1'
    assert config['OPTIONS']['PARSER_CLASS'] == 'redis.connection.HiredisParser'


def test_hiredis_socket_with_username_password_dj4(django_cache_url_dj4):
    url = 'hiredis://foo:bar@/path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'unix://foo@/path/to/socket?db=1&password=bar'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
    assert 'PARSER_CLASS' not in config.get('OPTIONS', {})
    assert config['KEY_PREFIX'] == 'site1'
    assert config['OPTIONS']['parser_class'] == 'redis.connection.HiredisParser'


def test_hiredis_with_password_dj3(django_cache_url_dj3):
    url = 'hiredis://:redispass@127.0.0.1:6379/0'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://127.0.0.1:6379/0'
    assert config['OPTIONS']['PASSWORD'] == 'redispass'
    assert config['OPTIONS']['PARSER_CLASS'] == 'redis.connection.HiredisParser'


def test_hiredis_with_password_dj4(django_cache_url_dj4):
    url = 'hiredis://:redispass@127.0.0.1:6379/0'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://:redispass@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
    assert 'PARSER_CLASS' not in config.get('OPTIONS', {})
    assert config['OPTIONS']['parser_class'] == 'redis.connection.HiredisParser'


def test_hiredis_with_username_dj3(django_cache_url_dj3):
    url = 'hiredis://foo:@127.0.0.1:6379/0'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://foo@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
    assert config['OPTIONS']['PARSER_CLASS'] == 'redis.connection.HiredisParser'


def test_hiredis_with_username_dj4(django_cache_url_dj4):
    url = 'hiredis://foo:@127.0.0.1:6379/0'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://foo:@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
    assert 'PARSER_CLASS' not in config.get('OPTIONS', {})
    assert config['OPTIONS']['parser_class'] == 'redis.connection.HiredisParser'


def test_hiredis_with_username_password_dj3(django_cache_url_dj3):
    url = 'hiredis://foo:bar@127.0.0.1:6379/0'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://foo@127.0.0.1:6379/0'
    assert config['OPTIONS']['PASSWORD'] == 'bar'
    assert config['OPTIONS']['PARSER_CLASS'] == 'redis.connection.HiredisParser'


def test_hiredis_with_username_password_dj4(django_cache_url_dj4):
    url = 'hiredis://foo:bar@127.0.0.1:6379/0'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://foo:bar@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
    assert 'PARSER_CLASS' not in config.get('OPTIONS', {})
    assert config['OPTIONS']['parser_class'] == 'redis.connection.HiredisParser'
# endregion


# region REDIS url
def test_redis_dj3(django_cache_url_dj3):
    url = 'redis://127.0.0.1:6379/0?key_prefix=site1'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DJANGO_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://127.0.0.1:6379/0'
    assert config['KEY_PREFIX'] == 'site1'


def test_redis_dj4(django_cache_url_dj4):
    url = 'redis://127.0.0.1:6379/0?key_prefix=site1'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://127.0.0.1:6379/0'
    assert config['KEY_PREFIX'] == 'site1'


def test_redis_with_password_dj3(django_cache_url_dj3):
    url = 'redis://:redispass@127.0.0.1:6379/0'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://127.0.0.1:6379/0'
    assert config['OPTIONS']['PASSWORD'] == 'redispass'


def test_redis_with_password_dj4(django_cache_url_dj4):
    url = 'redis://:redispass@127.0.0.1:6379/0'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://:redispass@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})


def test_redis_with_username_dj3(django_cache_url_dj3):
    url = 'redis://foo:@127.0.0.1:6379/0'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://foo@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})


def test_redis_with_username_dj4(django_cache_url_dj4):
    url = 'redis://foo:@127.0.0.1:6379/0'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://foo:@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})


def test_redis_with_username_password_dj3(django_cache_url_dj3):
    url = 'redis://foo:bar@127.0.0.1:6379/0'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DEFAULT_REDIS_BACKEND
    assert config['LOCATION'] == 'redis://foo@127.0.0.1:6379/0'
    assert config['OPTIONS']['PASSWORD'] == 'bar'


def test_redis_with_username_password_dj4(django_cache_url_dj4):
    url = 'redis://foo:bar@127.0.0.1:6379/0'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'redis://foo:bar@127.0.0.1:6379/0'
    assert 'PASSWORD' not in config.get('OPTIONS', {})
# endregion


# region REDIS socket
def test_redis_socket_dj3(django_cache_url_dj3):
    url = 'redis:///path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DJANGO_REDIS_BACKEND
    assert config['LOCATION'] == 'unix:///path/to/socket?db=1'
    assert 'OPTIONS' not in config
    assert config['KEY_PREFIX'] == 'site1'


def test_redis_socket_dj4(django_cache_url_dj4):
    url = 'redis:///path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'unix:///path/to/socket?db=1'
    assert 'OPTIONS' not in config
    assert config['KEY_PREFIX'] == 'site1'


def test_redis_socket_with_username_password_dj3(django_cache_url_dj3):
    url = 'redis://foo:bar@/path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DJANGO_REDIS_BACKEND
    assert config['LOCATION'] == 'unix://foo:bar@/path/to/socket?db=1'
    assert 'OPTIONS' not in config
    assert config['KEY_PREFIX'] == 'site1'


def test_redis_socket_with_username_password_dj4(django_cache_url_dj4):
    url = 'redis://foo:bar@/path/to/socket/1?key_prefix=site1'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'unix://foo@/path/to/socket?db=1&password=bar'
    assert 'OPTIONS' not in config
    assert config['KEY_PREFIX'] == 'site1'
# endregion


# region REDISS
def test_rediss_dj3(django_cache_url_dj3):
    url = 'rediss://127.0.0.1:6379'
    config = django_cache_url_dj3.parse(url)

    assert config['BACKEND'] == django_cache_url_dj3.DJANGO_REDIS_BACKEND
    assert config['LOCATION'] == 'rediss://127.0.0.1:6379/0'


def test_rediss_dj4(django_cache_url_dj4):
    url = 'rediss://127.0.0.1:6379'
    config = django_cache_url_dj4.parse(url)

    assert config['BACKEND'] == django_cache_url_dj4.BUILTIN_DJANGO_BACKEND
    assert config['LOCATION'] == 'rediss://127.0.0.1:6379/0'
# endregion


django_cache_url.py
# -*- coding: utf-8 -*-
import os
import re

try:
    import urllib.parse as urlparse
except ImportError:  # python 2
    import urlparse


# Register cache schemes in URLs.
urlparse.uses_netloc.append('db')
urlparse.uses_netloc.append('dummy')
urlparse.uses_netloc.append('file')
urlparse.uses_netloc.append('locmem')
urlparse.uses_netloc.append('uwsgicache')
urlparse.uses_netloc.append('memcached')
urlparse.uses_netloc.append('elasticache')
urlparse.uses_netloc.append('djangopylibmc')
urlparse.uses_netloc.append('pymemcached')
urlparse.uses_netloc.append('redis')
urlparse.uses_netloc.append('hiredis')

DEFAULT_ENV = 'CACHE_URL'

BACKENDS = {
    'db': 'django.core.cache.backends.db.DatabaseCache',
    'dummy': 'django.core.cache.backends.dummy.DummyCache',
    'elasticache': 'django_elasticache.memcached.ElastiCache',
    'file': 'django.core.cache.backends.filebased.FileBasedCache',
    'locmem': 'django.core.cache.backends.locmem.LocMemCache',
    'uwsgicache': 'uwsgicache.UWSGICache',
    'memcached': 'django.core.cache.backends.memcached.PyLibMCCache',
    'djangopylibmc': 'django_pylibmc.memcached.PyLibMCCache',
    'pymemcached': 'django.core.cache.backends.memcached.MemcachedCache',
    'redis': 'django_redis.cache.RedisCache',
    'hiredis': 'django_redis.cache.RedisCache',
}


def config(env=DEFAULT_ENV, default='locmem://'):
    """Returns configured CACHES dictionary from CACHE_URL"""
    config = {}

    s = os.environ.get(env, default)

    if s:
        config = parse(s)

    return config


def parse(url):
    """Parses a cache URL."""
    config = {}

    url = urlparse.urlparse(url)
    # Handle python 2.6 broken url parsing
    path, query = url.path, url.query
    if '?' in path and query == '':
        path, query = path.split('?', 1)

    cache_args = dict([(key.upper(), ';'.join(val)) for key, val in
                       urlparse.parse_qs(query).items()])

    # Update with environment configuration.
    backend = BACKENDS.get(url.scheme)
    if not backend:
        raise Exception('Unknown backend: "{0}"'.format(url.scheme))

    config['BACKEND'] = BACKENDS[url.scheme]

    redis_options = {}
    if url.scheme == 'hiredis':
        redis_options['PARSER_CLASS'] = 'redis.connection.HiredisParser'

    for key in ['CLIENT_CLASS', 'IGNORE_EXCEPTIONS']:
        val = cache_args.pop(key, None)
        if val is not None:
            redis_options[key] = val

    # File based
    if not url.netloc:
        if url.scheme in ('memcached', 'pymemcached', 'djangopylibmc'):
            config['LOCATION'] = 'unix:' + path

        elif url.scheme in ('redis', 'hiredis'):
            match = re.match(r'.+?(?P<db>\d+)', path)
            if match:
                db = match.group('db')
                path = path[:path.rfind('/')]
            else:
                db = '0'
            config['LOCATION'] = 'unix:%s?db=%s' % (path, db)
        else:
            config['LOCATION'] = path
    # URL based
    else:
        # Handle multiple hosts
        config['LOCATION'] = ';'.join(url.netloc.split(','))

        if url.scheme in ('redis', 'hiredis'):
            if url.password:
                redis_options['PASSWORD'] = url.password
            # Specifying the database is optional, use db 0 if not specified.
            db = path[1:] or '0'
            port = url.port if url.port else 6379
            config['LOCATION'] = "redis://%s:%s/%s" % (url.hostname, port, db)

    if redis_options:
        config.setdefault('OPTIONS', {}).update(redis_options)

    if url.scheme == 'uwsgicache':
        config['LOCATION'] = config.get('LOCATION', 'default') or 'default'

    # Pop special options from cache_args
    # https://docs.djangoproject.com/en/1.10/topics/cache/#cache-arguments
    options = {}
    for key in ['MAX_ENTRIES', 'CULL_FREQUENCY']:
        val = cache_args.pop(key, None)
        if val is not None:
            options[key] = int(val)
    if options:
        config.setdefault('OPTIONS', {}).update(options)

    config.update(cache_args)

    return config


django_cache_url/__init__.py
# -*- coding: utf-8 -*-
import os
import re

from django import VERSION

try:
    import urllib.parse as urlparse
except ImportError:  # python 2
    import urlparse


# Register cache schemes in URLs.
urlparse.uses_netloc.append('db')
urlparse.uses_netloc.append('dummy')
urlparse.uses_netloc.append('file')
urlparse.uses_netloc.append('locmem')
urlparse.uses_netloc.append('uwsgicache')
urlparse.uses_netloc.append('memcached')
urlparse.uses_netloc.append('elasticache')
urlparse.uses_netloc.append('djangopylibmc')
urlparse.uses_netloc.append('pymemcached')
urlparse.uses_netloc.append('pymemcache')
urlparse.uses_netloc.append('redis')
urlparse.uses_netloc.append('hiredis')

DEFAULT_ENV = 'CACHE_URL'
# TODO Remove as soon as Django 3.2 goes EOL
BUILTIN_DJANGO_BACKEND = 'django.core.cache.backends.redis.RedisCache'

DJANGO_REDIS_BACKEND = 'django_redis.cache.RedisCache'

DJANGO_REDIS_CACHE_LIB_KEY = 'redis-cache'
DJANGO_REDIS_CACHE_BACKEND = 'redis_cache.RedisCache'

DEFAULT_REDIS_BACKEND = DJANGO_REDIS_BACKEND if VERSION[0] < 4 else BUILTIN_DJANGO_BACKEND

BACKENDS = {
    'db': 'django.core.cache.backends.db.DatabaseCache',
    'dummy': 'django.core.cache.backends.dummy.DummyCache',
    'elasticache': 'django_elasticache.memcached.ElastiCache',
    'file': 'django.core.cache.backends.filebased.FileBasedCache',
    'locmem': 'django.core.cache.backends.locmem.LocMemCache',
    'uwsgicache': 'uwsgicache.UWSGICache',
    'memcached': 'django.core.cache.backends.memcached.PyLibMCCache',
    'djangopylibmc': 'django_pylibmc.memcached.PyLibMCCache',
    'pymemcached': 'django.core.cache.backends.memcached.MemcachedCache',
    'pymemcache': 'django.core.cache.backends.memcached.PyMemcacheCache',
    DJANGO_REDIS_CACHE_LIB_KEY: DJANGO_REDIS_CACHE_BACKEND,
    'redis': DEFAULT_REDIS_BACKEND,
    'rediss': DEFAULT_REDIS_BACKEND,
    'hiredis': DEFAULT_REDIS_BACKEND,
}


# django-redis-cache
def config(env=DEFAULT_ENV, default='locmem://'):
    """Returns configured CACHES dictionary from CACHE_URL"""
    config = {}

    s = os.environ.get(env, default)

    if s:
        config = parse(s)

    return config


def parse(url):
    """Parses a cache URL."""
    config = {}

    url = urlparse.urlparse(url)
    path, query = url.path, url.query

    cache_args = dict([(key.upper(), ';'.join(val)) for key, val in urlparse.parse_qs(query).items()])

    # Update with environment configuration.
    backend = BACKENDS.get(url.scheme)
    if not backend:
        raise Exception(f'Unknown backend: "{url.scheme}"')

    lib = cache_args.get('LIB')
    if 'redis' in url.scheme and lib in BACKENDS:
        backend = BACKENDS[lib]

    config['BACKEND'] = backend

    redis_options = {}
    if url.scheme == 'hiredis':
        redis_options['PARSER_CLASS'] = 'redis.connection.HiredisParser'

    # File based
    if url.hostname is None:  # only path exists
        if url.scheme in ('memcached', 'pymemcached', 'pymemcache', 'djangopylibmc'):
            config['LOCATION'] = 'unix://' + path

        elif url.scheme in ('redis', 'hiredis'):
            match = re.match(r'.+?(?P<db>\d+)', path)
            if match:
                db = match.group('db')
                path = path[:path.rfind('/')]
            else:
                db = '0'

            config['LOCATION'] = f'unix://{path}?db={db}'  # No auth

            username = url.username or ''
            password = url.password or ''
            if backend == DJANGO_REDIS_BACKEND:
                # django-redis socket connection:
                # unix://[[username]:[password]]@/path/to/socket.sock?db=0
                if username != '' or password != '':
                    config['LOCATION'] = f'unix://{username}:{password}@{path}?db={db}'
            elif backend == BUILTIN_DJANGO_BACKEND:
                # redis-py (native django 4 backend) socket connection:
                # unix://[username@]/path/to/socket.sock?db=0[&password=password]
                if username != '':
                    config['LOCATION'] = f'unix://{username}@{path}?db={db}'
                if password != '':
                    config['LOCATION'] += f'&password={password}'

                redis_options = parser_buildin_django_backend_options(cache_args, redis_options)

            elif backend == DJANGO_REDIS_CACHE_BACKEND:
                # django-redis-cache socket connection:
                # unix://[:password]@/path/to/socket.sock?db=0
                if username != '':
                    raise Exception('Username is not supported for unix socket connection with lib: redis-cache')
                if password != '':
                    config['LOCATION'] = f'unix://:{password}@{path}?db={db}'
        else:
            config['LOCATION'] = path
    # URL based
    else:
        # Handle multiple hosts
        config['LOCATION'] = ';'.join(url.netloc.split(','))

        # Redis
        if url.scheme in ('redis', 'rediss', 'hiredis'):
            # Specifying the database is optional, use db 0 if not specified.
            db = path[1:] or '0'
            port = url.port if url.port else 6379
            scheme = 'rediss' if url.scheme == 'rediss' else 'redis'
            config['LOCATION'] = f'{scheme}://{url.hostname}:{port}/{db}'  # No auth

            username = url.username or ''
            password = url.password or ''

            if backend == DJANGO_REDIS_BACKEND:
                # django-redis:
                # redis://[[username]:[password]]@localhost:6379/0
                # as for the password: `OPTIONS` does not overwrite the password in the `LOCATION`
                # but we still set password in `OPTIONS` rather than `LOCATION`
                if username != '':
                    config['LOCATION'] = f'{scheme}://{username}@{url.hostname}:{port}/{db}'

                if password != '':
                    redis_options['PASSWORD'] = password

            elif backend == BUILTIN_DJANGO_BACKEND:
                # redis-py (native django 4 backend):
                # redis://[[username]:[password]]@localhost:6379/0
                if username != '' or password != '':
                    config['LOCATION'] = f'{scheme}://{username}:{password}@{url.hostname}:{port}/{db}'

                redis_options = parser_buildin_django_backend_options(cache_args, redis_options)

            elif backend == DJANGO_REDIS_CACHE_BACKEND:
                # django-redis-cache:
                # redis://[:password]@localhost:6379/0
                if username != '':
                    raise Exception('Username is not supported for redis connection with lib: redis-cache')
                if password != '':
                    config['LOCATION'] = f'{scheme}://:{password}@{url.hostname}:{port}/{db}'

                # Update redis options
                if 'PARSER_CLASS' in cache_args:
                    redis_options['PARSER_CLASS'] = cache_args['PARSER_CLASS']

                if 'CONNECTION_POOL_CLASS' in cache_args:
                    redis_options['CONNECTION_POOL_CLASS'] = cache_args['CONNECTION_POOL_CLASS']

                if 'MAX_CONNECTIONS' in cache_args or 'TIMEOUT' in cache_args:
                    redis_options['CONNECTION_POOL_CLASS_KWARGS'] = {}

                    if 'MAX_CONNECTIONS' in cache_args:
                        redis_options['CONNECTION_POOL_CLASS_KWARGS']['max_connections'] = int(
                            cache_args['MAX_CONNECTIONS'])

                    if 'TIMEOUT' in cache_args:
                        redis_options['CONNECTION_POOL_CLASS_KWARGS']['timeout'] = int(cache_args['TIMEOUT'])

    if redis_options:
        config.setdefault('OPTIONS', {}).update(redis_options)

    if url.scheme == 'uwsgicache':
        config['LOCATION'] = config.get('LOCATION', 'default') or 'default'

    # Pop special options from cache_args
    # https://docs.djangoproject.com/en/1.10/topics/cache/#cache-arguments
    options = {}
    for key in ['MAX_ENTRIES', 'CULL_FREQUENCY']:
        val = cache_args.pop(key, None)
        if val is not None:
            options[key] = int(val)
    if options:
        config.setdefault('OPTIONS', {}).update(options)

    config.update(cache_args)

    return config


def parser_buildin_django_backend_options(cache_args, redis_options):
    if 'PARSER_CLASS' in cache_args:
        redis_options['PARSER_CLASS'] = cache_args['PARSER_CLASS']
    if 'POOL_CLASS' in cache_args:
        redis_options['POOL_CLASS'] = cache_args['POOL_CLASS']
    # native redis backend requires the keys in the `OPTIONS` to be lowercase
    redis_options = {k.lower(): v for k, v in redis_options.items()}
    return redis_options
